enhancements/ingress/client-tls.md

```diff
@@ -0,0 +1,315 @@
+functionality for configuring client TLS already exists in the template and can
+be configured using well defined environment variables.  The client certificate
+policy can be configured using the `ROUTER_MUTUAL_TLS_AUTH` variable.  The path
+to a CA certificate bundle can be specified using the
+`ROUTER_MUTUAL_TLS_AUTH_CA` variable, and similarly, the path to a file
+containing associated CRLs can be specified using the
+`ROUTER_MUTUAL_TLS_AUTH_CRL` variable.  A pattern for filtering client
+certificates by their subjects can be specified using the
+`ROUTER_MUTUAL_TLS_AUTH_FILTER` variable.  The ingress operator uses these
+environment variables to configure client TLS.
+
+To inject the CA certificate bundle into the router pod, the operator adds a
+configmap volume and volume mount and specifies the mount path using
+`ROUTER_MUTUAL_TLS_AUTH_CA` in the router deployment.
+
+To inject the CRL file into the router pod, the operator must first scan the
+certificates in the IngressController's CA certificate bundle configmap to
+determine whether any certificates specify any CRL destination endpoints.  If
+none does, then no CRL file is needed, and so no further configuration is
+required for CRLs.  However, if any certificate does specify any CRL
+distribution points, then the operator configures an additional configmap volume
+and volume mount for the CRLs and specifies the mount path using
+`ROUTER_MUTUAL_TLS_AUTH_CRL`.  The operator configures the deployment so that
+its pods cannot start until the CRL configmap exists.  This is necessary to
+ensure that revoked certificates are not permitted while the CRL is downloaded.
+
+A separate controller in the operator scans client CA certificate bundle
+configmaps, and if it finds that any certificates in a configmap specify any CRL
+destination points, then the controller downloads the CRLs and creates or
+updates the CRL configmap for the associated IngressController.
+
+Finally, if `spec.clientTLS.allowedSubjectPatterns` is nonempty, the operator
+combines the patterns therein and sets the result as the value for
+`ROUTER_MUTUAL_TLS_AUTH_FILTER` in the router deployment.
+
+### Risks and Mitigations
+
+If the underlying IngressController implementation were to change away from
+HAProxy to a different implementation, we would need to ensure that the new
+implementation supported the same capabilities.
+
+If a client CA certificate specifies a CRL distribution point that is invalid or
+inaccessible, and does not specify a valid and accessible distribution point,
+then the IngressController cannot start.  This is a deliberate design choice to
+avoid potentially allowing revoked client certificates.
+
+A CRL's refresh period may be determined from the CRL itself or from the HTTP
+response headers when the operator downloads a CRL, or the operator could simply
+refresh the CRL at a fixed period (for example, daily or hourly).  It is not
```

Comment:
If the CRL were to be deleted by a cluster administrator, would it be re-downloaded immediately? How would the Ingress Controller react to that? I could see a fixed refresh period working if a cluster administrator has the ability to manually "refresh" the CRL (via deletion, or otherwise) in the event of a quickly needed CRL update.

Reply:
Right, the operator redownloads the CRLs and recreates the configmap.  The kubelet updates the volume once the configmap is recreated (it does **not** update the volume while the configmap is absent, so the previous CRLs remain in effect in the meantime; [source](https://github.com/kubernetes/kubernetes/blob/4ad3018ac3f42d268ffbb84074fb1676d26bae90/pkg/volume/configmap/configmap.go#L197-L201)).